pandas/core/groupby.py

```diff
@@ -4437,20 +4437,15 @@ def _reorder_by_uniques(uniques, labels):
 
 
 def _groupby_indices(values):
+    """
+    return a dictionary of groups -> indices
+    """
+    if not is_categorical_dtype(values):
+        values = Categorical(values)
```

Comment:
I might be missing something, but wouldn't it possible to re-use the possibly existing factorization here?
```
def _groupby_indices(codes : Grouping.labels,  cats : Grouping.group_index)
```